# Assignment 5(a)

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

Q1 

Read the csv file credit_card_default into a dataframe. 

Display the first five rows of the dataset, display the last five rows of the dataset, and use the describe() function to output descriptive statistics.

In [2]:
df=pd.read_csv('credit_card_default.csv')
df.head()

ID  LIMIT_BAL  GENDER  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0   1      20000       2          2         1   24      2      2     -1   
1   2     120000       2          2         2   26     -1      2      0   
2   3      90000       2          2         2   34      0      0      0   
3   4      50000       2          2         1   37      0      0      0   
4   5      50000       1          2         1   57     -1      0     -1   

   PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0     -1  ...          0          0          0         0       689         0   
1      0  ...       3272       3455       3261         0      1000      1000   
2      0  ...      14331      14948      15549      1518      1500      1000   
3      0  ...      28314      28959      29547      2000      2019      1200   
4      0  ...      20940      19146      19131      2000     36681     10000   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
0         0         0         0        1  
1      1000         0      2000        1  
2      1000      1000      5000        0  
3      1100      1069      1000        0  
4      9000       689       679        0  

[5 rows x 25 columns]

In [3]:
df.tail()

ID  LIMIT_BAL  GENDER  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  \
29995  29996     220000       1          3         1   39      0      0   
29996  29997     150000       1          3         2   43     -1     -1   
29997  29998      30000       1          2         2   37      4      3   
29998  29999      80000       1          3         1   41      1     -1   
29999  30000      50000       1          2         1   46      0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
29995      0      0  ...      88004      31237      15980      8500     20000   
29996     -1     -1  ...       8979       5190          0      1837      3526   
29997      2     -1  ...      20878      20582      19357         0         0   
29998      0      0  ...      52774      11855      48944     85900      3409   
29999      0      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
29995      5003      3047      5000      1000        0  
29996      8998       129         0         0        0  
29997     22000      4200      2000      3100        1  
29998      1178      1926     52964      1804        1  
29999      1430      1000      1000      1000        1  

[5 rows x 25 columns]

In [4]:
df.describe()

ID       LIMIT_BAL        GENDER     EDUCATION      MARRIAGE  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603733      1.853133      1.551867   
std     8660.398374   129747.661567      0.489129      0.790349      0.521970   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                AGE         PAY_0         PAY_2         PAY_3         PAY_4  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      35.485500     -0.016700     -0.133767     -0.166200     -0.220667   
std        9.217904      1.123802      1.197186      1.196868      1.169139   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...      BILL_AMT4      BILL_AMT5      BILL_AMT6       PAY_AMT1  \
count  ...   30000.000000   30000.000000   30000.000000   30000.000000   
mean   ...   43262.948967   40311.400967   38871.760400    5663.580500   
std    ...   64332.856134   60797.155770   59554.107537   16563.280354   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2326.750000    1763.000000    1256.000000    1000.000000   
50%    ...   19052.000000   18104.500000   17071.000000    2100.000000   
75%    ...   54506.000000   50190.500000   49198.250000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

           PAY_AMT2      PAY_AMT3       PAY_AMT4       PAY_AMT5  \
count  3.000000e+04   30000.00000   30000.000000   30000.000000   
mean   5.921163e+03    5225.68150    4826.076867    4799.387633   
std    2.304087e+04   17606.96147   15666.159744   15278.305679   
min    0.000000e+00       0.00000       0.000000       0.000000   
25%    8.330000e+02     390.00000     296.000000     252.500000   
50%    2.009000e+03    1800.00000    1500.000000    1500.000000   
75%    5.000000e+03    4505.00000    4013.250000    4031.500000   
max    1.684259e+06  896040.00000  621000.000000  426529.000000   

            PAY_AMT6       Default  
count   30000.000000  30000.000000  
mean     5215.502567      0.221200  
std     17777.465775      0.415062  
min         0.000000      0.000000  
25%       117.750000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]

Q2

Set X equal to the dataframe subset that contains the ‘LIMIT_BAL’ and ‘AGE’ columns and includes all the columns from ‘PAY_0’ to ‘PAY_AMT6’. 

In this selection, we are only selecting numerical data and not categorical data. 

Display the first five rows of this subset dataframe.

In [5]:
X_pay = df.loc[:,'PAY_0':'PAY_AMT6']
X_pay

PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  \
0          2      2     -1     -1     -2     -2       3913       3102   
1         -1      2      0      0      0      2       2682       1725   
2          0      0      0      0      0      0      29239      14027   
3          0      0      0      0      0      0      46990      48233   
4         -1      0     -1      0      0      0       8617       5670   
...      ...    ...    ...    ...    ...    ...        ...        ...   
29995      0      0      0      0      0      0     188948     192815   
29996     -1     -1     -1     -1      0      0       1683       1828   
29997      4      3      2     -1      0      0       3565       3356   
29998      1     -1      0      0      0     -1      -1645      78379   
29999      0      0      0      0      0      0      47929      48905   

       BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0            689          0          0          0         0       689   
1           2682       3272       3455       3261         0      1000   
2          13559      14331      14948      15549      1518      1500   
3          49291      28314      28959      29547      2000      2019   
4          35835      20940      19146      19131      2000     36681   
...          ...        ...        ...        ...       ...       ...   
29995     208365      88004      31237      15980      8500     20000   
29996       3502       8979       5190          0      1837      3526   
29997       2758      20878      20582      19357         0         0   
29998      76304      52774      11855      48944     85900      3409   
29999      49764      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0             0         0         0         0  
1          1000      1000         0      2000  
2          1000      1000      1000      5000  
3          1200      1100      1069      1000  
4         10000      9000       689       679  
...         ...       ...       ...       ...  
29995      5003      3047      5000      1000  
29996      8998       129         0         0  
29997     22000      4200      2000      3100  
29998      1178      1926     52964      1804  
29999      1430      1000      1000      1000  

[30000 rows x 18 columns]

In [6]:
X_num = pd.concat((df[['LIMIT_BAL',"AGE"]],X_pay),axis =1)
X_num

LIMIT_BAL  AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  \
0          20000   24      2      2     -1     -1     -2     -2       3913   
1         120000   26     -1      2      0      0      0      2       2682   
2          90000   34      0      0      0      0      0      0      29239   
3          50000   37      0      0      0      0      0      0      46990   
4          50000   57     -1      0     -1      0      0      0       8617   
...          ...  ...    ...    ...    ...    ...    ...    ...        ...   
29995     220000   39      0      0      0      0      0      0     188948   
29996     150000   43     -1     -1     -1     -1      0      0       1683   
29997      30000   37      4      3      2     -1      0      0       3565   
29998      80000   41      1     -1      0      0      0     -1      -1645   
29999      50000   46      0      0      0      0      0      0      47929   

       BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  \
0           3102        689          0          0          0         0   
1           1725       2682       3272       3455       3261         0   
2          14027      13559      14331      14948      15549      1518   
3          48233      49291      28314      28959      29547      2000   
4           5670      35835      20940      19146      19131      2000   
...          ...        ...        ...        ...        ...       ...   
29995     192815     208365      88004      31237      15980      8500   
29996       1828       3502       8979       5190          0      1837   
29997       3356       2758      20878      20582      19357         0   
29998      78379      76304      52774      11855      48944     85900   
29999      48905      49764      36535      32428      15313      2078   

       PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0           689         0         0         0         0  
1          1000      1000      1000         0      2000  
2          1500      1000      1000      1000      5000  
3          2019      1200      1100      1069      1000  
4         36681     10000      9000       689       679  
...         ...       ...       ...       ...       ...  
29995     20000      5003      3047      5000      1000  
29996      3526      8998       129         0         0  
29997         0     22000      4200      2000      3100  
29998      3409      1178      1926     52964      1804  
29999      1800      1430      1000      1000      1000  

[30000 rows x 20 columns]

In [7]:
# The other way to do so
X = df.copy()

In [8]:
X = X.drop(['ID', 'GENDER','EDUCATION','MARRIAGE','Default'], axis=1)

In [9]:
X.head()

LIMIT_BAL  AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  \
0      20000   24      2      2     -1     -1     -2     -2       3913   
1     120000   26     -1      2      0      0      0      2       2682   
2      90000   34      0      0      0      0      0      0      29239   
3      50000   37      0      0      0      0      0      0      46990   
4      50000   57     -1      0     -1      0      0      0       8617   

   BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0       3102        689          0          0          0         0       689   
1       1725       2682       3272       3455       3261         0      1000   
2      14027      13559      14331      14948      15549      1518      1500   
3      48233      49291      28314      28959      29547      2000      2019   
4       5670      35835      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0         0         0         0         0  
1      1000      1000         0      2000  
2      1000      1000      1000      5000  
3      1200      1100      1069      1000  
4     10000      9000       689       679

Q3

Set y equal to the ‘Default’ column.

In [10]:
y = df['Default'].copy()

In [11]:
y.head()

0    1
1    1
2    0
3    0
4    0
Name: Default, dtype: int64

In [12]:
type(y)

pandas.core.series.Series

Q4

Split the data into a training and a test dataset using the following command:

from sklearn.model_selection import train_test_split

Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, random_state=0

In [13]:
from sklearn.model_selection import train_test_split
# split the data with 50% in each set
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, random_state=0)

Q5 

Run Gaussian Naive Bayes to classify labels into default (1) or not default (0).

In [14]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(Xtrain, ytrain)
y_model = model.predict(Xtest)

Q6 

Calculate the accuracy score. Provide a one line comment to explain why the results are not so good.

In [15]:
from sklearn.metrics import accuracy_score
accuracy_score(ytest, y_model)

0.36093333333333333

comment: the model doesn't seem to do a good job since the score is low, merely 0.361

Q7

Extract two PCA factors from the payment data subset containing the following 18 columns only: ‘PAY_0’ to 'PAY_AMT6. Concatenate these two PCA factors with the following columns from the original dataframe: ‘LIMIT_BAL’ and ‘AGE’ columns.

In [16]:
X_subset = X.drop(['LIMIT_BAL','AGE'],axis = 1).copy()
X_subset

PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  \
0          2      2     -1     -1     -2     -2       3913       3102   
1         -1      2      0      0      0      2       2682       1725   
2          0      0      0      0      0      0      29239      14027   
3          0      0      0      0      0      0      46990      48233   
4         -1      0     -1      0      0      0       8617       5670   
...      ...    ...    ...    ...    ...    ...        ...        ...   
29995      0      0      0      0      0      0     188948     192815   
29996     -1     -1     -1     -1      0      0       1683       1828   
29997      4      3      2     -1      0      0       3565       3356   
29998      1     -1      0      0      0     -1      -1645      78379   
29999      0      0      0      0      0      0      47929      48905   

       BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0            689          0          0          0         0       689   
1           2682       3272       3455       3261         0      1000   
2          13559      14331      14948      15549      1518      1500   
3          49291      28314      28959      29547      2000      2019   
4          35835      20940      19146      19131      2000     36681   
...          ...        ...        ...        ...       ...       ...   
29995     208365      88004      31237      15980      8500     20000   
29996       3502       8979       5190          0      1837      3526   
29997       2758      20878      20582      19357         0         0   
29998      76304      52774      11855      48944     85900      3409   
29999      49764      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0             0         0         0         0  
1          1000      1000         0      2000  
2          1000      1000      1000      5000  
3          1200      1100      1069      1000  
4         10000      9000       689       679  
...         ...       ...       ...       ...  
29995      5003      3047      5000      1000  
29996      8998       129         0         0  
29997     22000      4200      2000      3100  
29998      1178      1926     52964      1804  
29999      1430      1000      1000      1000  

[30000 rows x 18 columns]

In [17]:
from sklearn.decomposition import PCA
model = PCA(n_components=2)
model.fit(X_subset)
X_subset_pca = model.transform(X_subset)

In [18]:
X_new = X[['LIMIT_BAL','AGE']].copy()
X_new['pca1'] = X_subset_pca[:,0].copy()
X_new['pca2'] = X_subset_pca[:,1].copy()
X_new

LIMIT_BAL  AGE           pca1           pca2
0          20000   24 -107827.409967   -3025.439727
1         120000   26 -104308.440907    2234.471996
2          90000   34  -69249.080179   -3943.555615
3          50000   37  -15282.889847  -12499.864892
4          50000   57  -65854.717174   22834.675395
...          ...  ...            ...            ...
29995     220000   39  200781.583886 -141245.646798
29996     150000   43 -102347.512303    4922.065594
29997      30000   37  -83565.143328   23864.164099
29998      80000   41    1280.529291   16497.463212
29999      50000   46  -14836.706495  -16559.321778

[30000 rows x 4 columns]

Q8

Re-run the Gaussian Naive Bayes algorithm to classify labels into default(1)or notd efault (0) but where now the features contain ‘LIMIT_BAL’, ‘AGE’ and the two payment PCA factors.

In [19]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X_new, y, random_state=0)

model_new = GaussianNB()
model_new.fit(Xtrain, ytrain)
y_model = model_new.predict(Xtest)

Q9 

Calculate the accuracy score. Provide a one line comment to explain why the results have improved.

In [20]:
accuracy_score(ytest, y_model)

0.7824

comment: we only replace ‘PAY_0’ to 'PAY_AMT6' by two PCA vectors, but the performance is much better than the case in the previous model

Q10 

Run the following code to transform the categorical data into a structure that we can use

data['GENDER'] = data['GENDER'].map({1: 1, 2: 0})

data = pd.get_dummies(data, columns=['EDUCATION','MARRIAGE']) 

data.head()

This code creates a series of dummy variable features for each label in a categorical feature.

In [21]:
df['GENDER'] = df['GENDER'].map({1: 1, 2: 0})
df = pd.get_dummies(df, columns=['EDUCATION','MARRIAGE'])
df.head()

ID  LIMIT_BAL  GENDER  AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  ...  \
0   1      20000       0   24      2      2     -1     -1     -2     -2  ...   
1   2     120000       0   26     -1      2      0      0      0      2  ...   
2   3      90000       0   34      0      0      0      0      0      0  ...   
3   4      50000       0   37      0      0      0      0      0      0  ...   
4   5      50000       1   57     -1      0     -1      0      0      0  ...   

   EDUCATION_1  EDUCATION_2  EDUCATION_3  EDUCATION_4  EDUCATION_5  \
0            0            1            0            0            0   
1            0            1            0            0            0   
2            0            1            0            0            0   
3            0            1            0            0            0   
4            0            1            0            0            0   

   EDUCATION_6  MARRIAGE_0  MARRIAGE_1  MARRIAGE_2  MARRIAGE_3  
0            0           0           1           0           0  
1            0           0           0           1           0  
2            0           0           0           1           0  
3            0           0           1           0           0  
4            0           0           1           0           0  

[5 rows x 34 columns]

Q11

Include these categorical dummy variable features in a dataframe along with the features: ‘LIMIT_BAL’, ‘AGE’ and the two payment PCA factors. Run the Gaussian Naive Bayes algorithm to classify labels into default (1) or not default (0) based on this extended set of features.


In [22]:
X_new =X_new.join(df[['EDUCATION_0','EDUCATION_1','EDUCATION_2','EDUCATION_3','EDUCATION_4'
              ,'EDUCATION_5','EDUCATION_6','MARRIAGE_0','MARRIAGE_1','MARRIAGE_2','MARRIAGE_3']])
X_new

LIMIT_BAL  AGE           pca1           pca2  EDUCATION_0  EDUCATION_1  \
0          20000   24 -107827.409967   -3025.439727            0            0   
1         120000   26 -104308.440907    2234.471996            0            0   
2          90000   34  -69249.080179   -3943.555615            0            0   
3          50000   37  -15282.889847  -12499.864892            0            0   
4          50000   57  -65854.717174   22834.675395            0            0   
...          ...  ...            ...            ...          ...          ...   
29995     220000   39  200781.583886 -141245.646798            0            0   
29996     150000   43 -102347.512303    4922.065594            0            0   
29997      30000   37  -83565.143328   23864.164099            0            0   
29998      80000   41    1280.529291   16497.463212            0            0   
29999      50000   46  -14836.706495  -16559.321778            0            0   

       EDUCATION_2  EDUCATION_3  EDUCATION_4  EDUCATION_5  EDUCATION_6  \
0                1            0            0            0            0   
1                1            0            0            0            0   
2                1            0            0            0            0   
3                1            0            0            0            0   
4                1            0            0            0            0   
...            ...          ...          ...          ...          ...   
29995            0            1            0            0            0   
29996            0            1            0            0            0   
29997            1            0            0            0            0   
29998            0            1            0            0            0   
29999            1            0            0            0            0   

       MARRIAGE_0  MARRIAGE_1  MARRIAGE_2  MARRIAGE_3  
0               0           1           0           0  
1               0           0           1           0  
2               0           0           1           0  
3               0           1           0           0  
4               0           1           0           0  
...           ...         ...         ...         ...  
29995           0           1           0           0  
29996           0           0           1           0  
29997           0           0           1           0  
29998           0           1           0           0  
29999           0           1           0           0  

[30000 rows x 15 columns]

In [23]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X_new, y, random_state=0)

model_new = GaussianNB()
model_new.fit(Xtrain, ytrain)
y_model = model_new.predict(Xtest)

In [24]:
accuracy_score(ytest, y_model)

0.7824

In [25]:
ytest.mean()

0.2176

The model in fact always predicts not default.